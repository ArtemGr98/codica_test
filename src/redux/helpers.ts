export const findIndexCity = (cities: { name: string }[], name: string) => {
    return cities.length ? cities.findIndex(city => city.name === name) : -1
}

export const getLocalStorageCities = () => {
    let cities: { name: string }[] = []

    if (localStorage.getItem('cities')) {
        cities = JSON.parse(localStorage.getItem('cities') as string)
    }
    return cities
}