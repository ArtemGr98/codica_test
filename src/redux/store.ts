import { configureStore } from '@reduxjs/toolkit'
import citiesReducer from './citiesSlice'

export const store = configureStore({
    reducer: {
        cities: citiesReducer
    },
})

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;

