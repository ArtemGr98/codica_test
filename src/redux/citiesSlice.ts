import {createAsyncThunk, createSlice, PayloadAction} from '@reduxjs/toolkit'
import {cityWeatherApi, CityWeatherResponseI} from '../api/cityWeatherApi'
import {RootState} from './store'
import {AxiosError} from 'axios'
import {findIndexCity, getLocalStorageCities} from './helpers'

export interface CitiesStateI {
    loading: boolean,
    error: undefined | string,
    apiKey: string,
    units: string,
    citiesCards: CityWeatherResponseI[]
}

const initialState: CitiesStateI = {
    loading: false,
    error: undefined,
    apiKey: '1b0a340168db93a5a6d8f9db271a81ed',
    units: 'metric',
    citiesCards: []
}

export const getCityWeatherAsync = createAsyncThunk<CityWeatherResponseI,
    string, { state: RootState, rejectValue: string }>(
        'cities/getCityWeatherAsync',
        async (city, {getState, rejectWithValue}) => {
            try {
                const {units, apiKey} = getState().cities
                const response = await cityWeatherApi.getCityWeather(city, units, apiKey)

                const cities = getLocalStorageCities()
                const isNewCity = findIndexCity(cities, response.name)

                if (isNewCity === -1) {
                    cities.push({name: response.name})
                    localStorage.setItem('cities', JSON.stringify(cities))
                }
                return response
            } catch (err) {
                const error: AxiosError<string> = err as any
                return rejectWithValue(error.message)
            }
        }
    )

export const citiesSlice = createSlice({
    name: 'cities',
    initialState,
    reducers: {
        removeCity: (state, action: PayloadAction<string>) => {
            const removeCityIndex = findIndexCity(state.citiesCards, action.payload)
            state.citiesCards.splice(removeCityIndex, 1)

            const cities = getLocalStorageCities()
            const removeLSIndex = findIndexCity(cities, action.payload)
            cities.splice(removeLSIndex, 1)
            cities.length ?
                localStorage.setItem('cities', JSON.stringify(cities)) :
                localStorage.removeItem('cities')
        }
    },
    extraReducers: (builder) => {
        builder
            .addCase(getCityWeatherAsync.pending, (state) => {
                state.loading = true
                state.error = undefined
            })
            .addCase(getCityWeatherAsync.fulfilled, (state, {payload}) => {
                state.loading = false

                const isNewCity = findIndexCity(state.citiesCards, payload.name)
                isNewCity === -1 ?
                    state.citiesCards.push({
                        name: payload.name,
                        main: {
                            temp: payload.main.temp,
                            humidity: payload.main.humidity,
                            pressure: payload.main.pressure,
                            feels_like: payload.main.feels_like
                        },
                        sys: {country: payload.sys.country},
                        weather: [{
                            main: payload.weather[0].main,
                            description: payload.weather[0].description
                        }]
                    }) :
                    state.citiesCards[isNewCity] = {
                        name: payload.name,
                        main: {
                            temp: payload.main.temp,
                            humidity: payload.main.humidity,
                            pressure: payload.main.pressure,
                            feels_like: payload.main.feels_like
                        },
                        sys: {country: payload.sys.country},
                        weather: [{
                            main: payload.weather[0].main,
                            description: payload.weather[0].description
                        }]
                    }
            })
            .addCase(getCityWeatherAsync.rejected, (state, action) => {
                state.loading = false
                state.error = action.payload
            })
    },
})




export const {removeCity} = citiesSlice.actions
export default citiesSlice.reducer
