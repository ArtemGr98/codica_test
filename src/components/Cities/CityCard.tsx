import {FC, useState} from 'react'
import {useAppDispatch} from '../../hooks/hooks'
import {getCityWeatherAsync, removeCity} from '../../redux/citiesSlice'
import {
    Box,
    Button,
    Card,
    CardActions,
    CardContent,
    Grid,
    IconButton,
    styled,
    Typography
} from '@mui/material'
import ExpandMoreIcon from '@mui/icons-material/ExpandMore'
import DeleteIcon from '@mui/icons-material/Delete'
import {CityCardWeatherI, ExpandMoreProps} from "../../types/types";

const ExpandMore = styled((props: ExpandMoreProps) => {
    const {expand, ...other} = props
    return <IconButton {...other} />
})(({theme, expand}) => ({
    transform: !expand ? 'rotate(0deg)' : 'rotate(180deg)',
    marginLeft: 'auto',
    transition: theme.transitions.create('transform', {
        duration: theme.transitions.duration.shortest,
    }),
}))

interface CityCardProps {
    cityCardWeather: CityCardWeatherI
}

const CityCard: FC<CityCardProps> = ({cityCardWeather}) => {
    const {
        name, temp, country, weather_main,
        weather_description, humidity, feels_like, pressure
    } = cityCardWeather

    const [expanded, setExpanded] = useState(false)
    const dispatch = useAppDispatch()

    return (
        <Grid item xs={6} sm={4} md={3} lg={2}>
            <Card>
                <CardContent>
                    <Typography variant="h6">
                        {name} {country}
                    </Typography>
                    <Typography>{Math.round(temp)} ℃</Typography>
                    {
                        expanded ? <Typography>{weather_description}</Typography> :
                            <Typography>{weather_main}</Typography>
                    }

                    {expanded && <div>
                        <Typography>humidity: {humidity} %</Typography>
                        <Typography>pressure: {pressure} mm</Typography>
                        <Typography>feels_like: {Math.round(feels_like)} ℃</Typography>
                    </div>}
                </CardContent>
                <CardActions>
                    <Box>
                        <Button onClick={() => dispatch(getCityWeatherAsync(name))}>update</Button>
                        <IconButton aria-label="delete" onClick={() => dispatch(removeCity(name))}>
                            <DeleteIcon/>
                        </IconButton>
                    </Box>

                    <ExpandMore
                        expand={expanded}
                        onClick={() => setExpanded(!expanded)}
                        aria-expanded={expanded}
                        aria-label="show more"
                    >
                        <ExpandMoreIcon/>
                    </ExpandMore>
                </CardActions>
            </Card>
        </Grid>
    )
}

export default CityCard