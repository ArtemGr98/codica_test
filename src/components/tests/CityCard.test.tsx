import {render, screen, fireEvent} from '@testing-library/react'
import CityCard, {CityWeatherDataI} from '../Cities/CityCard'
import {mockedDispatch} from './helpers'
import * as actions from './../../redux/citiesSlice'

const cityWeatherData: CityWeatherDataI = {
    name: 'city',
    temp: 2,
    country: 'UA',
    weather_main: 'snow',
    weather_description: 'snow',
    humidity: 98,
    pressure: 1000,
    feels_like: 0
}

describe('CityCard component', () => {
    it('CityCard render', () => {
        const view = render(<CityCard cityWeatherData={cityWeatherData}/>)

        expect(view).toMatchSnapshot()
    })
    it('should dispatch actions', () => {
        const dispatch = jest.fn()
        mockedDispatch.mockReturnValue(dispatch)
        const mockedRemoveCity = jest.spyOn(actions, 'removeCity')
        const mockedUpdateCity = jest.spyOn(actions, 'getCityWeatherAsync')

        render(<CityCard cityWeatherData={cityWeatherData}/>)

        fireEvent.click(screen.getByRole('button', {name: /delete/i}))
        expect(dispatch).toHaveBeenCalledTimes(1)
        expect(mockedRemoveCity).toHaveBeenCalledWith(cityWeatherData.name)

        fireEvent.click(screen.getByRole('button', {name: 'update'}))
        expect(dispatch).toHaveBeenCalledTimes(2)
        expect(mockedUpdateCity).toHaveBeenCalledWith(cityWeatherData.name)
    })
})