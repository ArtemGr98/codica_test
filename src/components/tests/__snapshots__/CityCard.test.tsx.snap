// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityCard component CityCard render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 MuiGrid-grid-lg-2 css-13i4rnv-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              city
               
              UA
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              2
               ℃
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              snow
            </p>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <div
              class="MuiBox-root css-0"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                update
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="delete"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <button
              aria-expanded="false"
              aria-label="show more"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-3ogxn6-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ExpandMoreIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 MuiGrid-grid-lg-2 css-13i4rnv-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            city
             
            UA
          </h6>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            2
             ℃
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            snow
          </p>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              update
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="delete"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <button
            aria-expanded="false"
            aria-label="show more"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-3ogxn6-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ExpandMoreIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
